### ISSUE TITLE
USWDS NPM refactor for single page applications

### ISSUE THREAD

**Issue by jAigret-Bix** [uswds/uswds#3742]
Update 11/25:
After further discussion we have determined that we need to finish the Pattern Lab folder structure first. Then we will determine how to best structure (and restructure) the distribution package. Dependencies are cleaned up. This issue is blocked by #3820.  We'll need to break this into smaller issues when we better understand next steps. 

_________________
**Description**
We will re-factor the current NPM package to be more lean and include entry points to each component for easier importation of modules in modern javascript frameworks without affecting the integrity of the development platform.

- [x] Clean up dependencies
- [ ] Finish the Pattern Lab folder structure
- [ ] Determine how to best structure (and restructure) the distribution package
- [ ] Update the Gulp build task

**DOD**
- [ ] Working version of the USWDS distribution package
- [ ] Create new issues for publishing NPM

**Comment by mejiaj** [comment]
@sarnikbixal not sure if this is the right place for it.

https://www.youtube.com/watch?v=cLxNdLK--yI

**Comment by thisisdano** [comment]
Let's talk about this one. I don't think I understand what needs to happen.

**Comment by aduth** [comment]
I was about to create a new issue before discovering this one, and I think it sufficiently overlaps. I'll just copy over the text I'd written up, in case it's useful for how I'd considered the problem:

---

**Is your feature request related to a problem? Please describe.**

As a user, I expect websites to load quickly, so that I can access the content I need without delay.

In recent versions, USWDS has added a number of new components with rich interactivity. This is great, but it comes at a cost: the size of the main JavaScript bundle has nearly tripled between versions 2.0.0 and 2.12.0, from [8kb](https://bundlephobia.com/package/uswds@2.0.0) to [23kb](https://bundlephobia.com/package/uswds@2.12.0). Much of this can be unnecessary as well, since code for every component is loaded, regardless of whether those components are actually present in the page. Due to how component implementations use [event delegation](https://www.npmjs.com/package/receptor), this also means that event listeners for all components are bound to the document regardless whether any instance of that component exists in the page (example of `mousemove` described in https://github.com/uswds/uswds/issues/4184#issuecomment-872394018).

**Describe the solution you'd like**

Ideally, only the scripts for components present in the page are loaded.

I could envision a couple options for solutions:

- **Implicitly**, the USWDS script detects which elements are on the page, and loads-on-demand only the scripts for components present.
   - Browser support is a concern, but semantically this is well-aligned to [dynamic imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#dynamic_imports). This can also be emulated in many bundlers through code splitting ([Webpack documentation](https://webpack.js.org/guides/code-splitting/)).
   - Pseudo-code: 
     ```js
     ['usa-file-input'].forEach((component) => document.querySelector(`.${component}`) && import(`./components/${component}`));
     ```
- **Explicitly**, a developer loads and (optionally) initializes the specific components that they plan to use.

With the "explicit" solution, I could imagine one or the other (or both!) of loading with or without side effects.

With side-effects, the intent is to "register" or "define" the component as being used, and USWDS would handle all behaviors, including for components added after page load (related comment at https://github.com/uswds/uswds/issues/2884#issuecomment-872238628):

```js
import 'uswds/accordion/define';
import 'uswds/input-prefix-suffix/define';
// ...
```

Without side-effects, USWDS would export either a class or initializer function for each component and delegate responsibility to the consuming project to initialize them as desired:

```js
import { Accordion } from 'uswds';

document.querySelectorAll('.usa-accordion').forEach((element) => {
  new Accordion(element).bind();
});
```

Related concepts:

- [Tree shaking](https://webpack.js.org/guides/tree-shaking/)
   - Also note [`package.json` `sideEffects` property](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free) would be relevant for the "side-effecty" example above, to differentiate between pure and side-effect entry-points (e.g. `import Accordion from 'uswds/accordion'` vs `import 'uswds/accordion/define'`).
- [`package.json` `exports` for multiple entry points](https://nodejs.org/api/packages.html#packages_package_entry_points)

**Comment by mejiaj** [comment]
Related PR #4624. This is safe to close.
